Combine effectors into a fully working 'addServiceToCluster' effector

tbouron opened the pull request:
This effector will add, install and starts an known service onto an existing cluster. It will block until the end of the entire operation and report back errors if any.

duncangrant commented on ambari/src/main/java/io/brooklyn/ambari/server/AmbariServerImpl.java:
```diff
@@ -224,63 +232,122 @@ public void addHostToCluster(@EffectorParam(name = "Cluster name") String cluste
+    public void addServiceToCluster(@EffectorParam(name = "Cluster name") final String cluster,
+                                    @EffectorParam(name = "Service name") final String service,
+                                    @EffectorParam(name = "Mappings of component to host") Map<String, String> mappings,
+                                    @EffectorParam(name = "Service configuration key", nullable = true, defaultValue = EffectorParam.MAGIC_STRING_MEANING_NULL) String configurationKey,
```
What's a service configuration key?

tbouron replied:
That's the key for the configuration map like `nagios-env` or `hdfs-site`. Here is [an example](https://github.com/brooklyncentral/brooklyn-ambari/blob/master/ambari/src/test/resources/ambari-cluster.yaml#L11).

There is one thing I realise though, the last 2 parameters could be merged into one big `Map<String, Map<String, String>>`. WDYT?

duncangrant replied:
What of there is more than one?  e.g. hbase-site and hbase-policy?

tbouron replied:
Yes, that's why I was suggesting to replace parameters `configurationKey` and `configuration` by only one parameter `Map<String, Map<String, String>> configuration`

duncangrant commented on ambari/src/main/java/io/brooklyn/ambari/server/AmbariServerImpl.java:
```diff
@@ -224,63 +232,122 @@ public void addHostToCluster(@EffectorParam(name = "Cluster name") String cluste
-        restAdapter.create(HostEndpoint.class).createHostComponent(cluster, hostName, component);
+
+        // Step 1 - Add the service to the cluster
+        restAdapter.create(ServiceEndpoint.class).addService(cluster, service);
```
For clarity I'd remove restAdapter.create(HostEndpoint.class) and replace with hostEndpoint or hostEndpoint()
Also for all occurrences below.

duncangrant commented on ambari/src/main/java/io/brooklyn/ambari/server/AmbariServerImpl.java:
```diff
@@ -224,63 +232,122 @@ public void addHostToCluster(@EffectorParam(name = "Cluster name") String cluste
+        });
+
+        try {
+            requestFuture.get(5, TimeUnit.MINUTES);
```
Why do we do this?
It looks like we are blocking a thread for no reason when we could just have a sensor for when the request finishes.

tbouron replied:
Yes, the thread is blocked because we need to wait for the operation to finish to report back. Not doing this means that as soon as we subscribe to the sensor, the effector finishes, reporting `COMPLETED` within the Brooklyn UI which we obviously don't want

duncangrant replied:
Is that because you haven't added it to an entity?

tbouron replied:
What? No.

`requestFuture.get(5, TimeUnit.MINUTES);` blocks until `requestFuture` has a value or exception set (which happens on line 310 or 314) or if it reaches the defined timeout (5 minutes) to avoid a possible infinite loop.

Does it make sense?

duncangrant commented:
Any chance we can write a live test for this?  It would help to document the interface as well as to make sure that this still works when we upgrade ambari versions.

tbouron commented:
@duncangrant I think I can write live tests for it. Do you want that for this PR or can I add it a bit later?

tbouron commented:
@duncangrant I addressed previous comments as well as:
- used sub-task mechanism for the installation / start part
- added unit tests
- added live tests

duncangrant commented on ambari/src/main/java/io/brooklyn/ambari/rest/RequestCheckRunnable.java:
```diff
@@ -0,0 +1,129 @@
+    @Override
+    public void run() {
+        boolean done = Repeater.create(String.format("Request %s status check", builder.request.toString()))
+                .every(Duration.ONE_SECOND)
```
Should we use exponential back-off here?

tbouron replied:
We could use exponential back-off but I'm not sure if it is the best thing to do as you might hit the timeout before the repeater retries to verify the result. Does it make sense?
